broadinstitute/single_cell_portal_core
# Client-side file validation in upload for gzipped files (SCP-4160)

This adds partial CSFV for files compressed via gzip.  It drastically improves time-to-fail on a large portion of uploaded files.

Previously, gzipped files had superficial validation, checking for a proper first byte and file extension.  Now, all CSFV rules are applied for up to 500 lines of an uploaded gzipped file, newly benefitting [~15% of uploaded files](https://mixpanel.com/s/4xDTGv).  Decompression is handled by the small but mighty [fflate](https://github.com/101arrowz/fflate) JavaScript library.

Here's how it looks:

https://user-images.githubusercontent.com/1334561/158390985-02d50413-2fce-4247-b913-8275f5de5a36.mov

Also included is preliminary refactoring for live reload using Vite.  This builds on #1397.  Live reload works for React components like `FileUploadControl.js`, but is still a work-in-progress for libraries like `validate-file.js`, `io.js`, etc.

To test:
* Go to upload page
* Upload an invalid gzipped file, e.g. [Dense_matrix_204_MB_bad_GENE_header.txt.gz](https://storage.googleapis.com/broad-singlecellportal-public/test/Dense_matrix_204_MB_bad_GENE_header.txt.gz)
* Confirm error message and warning message appear
* Upload a valid gzipped file, e.g. [Dense_matrix_204_MB_good.txt.gz](https://storage.googleapis.com/broad-singlecellportal-public/test/Dense_matrix_204_MB_good.txt.gz)
* Confirm error message does not appear, and warning message does appear

This satisfies SCP-4160.
